### PR TITLE
New detector: `EtaLikeSimplifications`

### LINKED ISSUE #92: Detect eta-like simplifications
**Summary**

Things that can be simplified using eta-rules (eta-conversion in type theory):
- booleans: `cond ? true : false` -> `cond`, for any boolean expression `cond`, or 
- structs: `Foo {x: ex, y: ey}.x` -> `ex`, for any expression `ex`.

are most likely produced either by an inexperienced programmer or result from a typo or other form of mistake.

This also can be applied to statements as well, see the example below.

**Context**

I don't think it's a high-priority detector, although I've seen code like this (especially in the example below) in the wild.

**Examples**

```tact
fun foo(x : Int): Bool {
  // ... some logic here
  if (condition) {
    return true;
  } else {
    return false;
  }
}
```

Use instead:

```tact
fun foo(x : Int): Bool {
  // ... some logic here
  return condition;
}
```

One more example with `==` for booleans:

```tact
fun foo(x : Int): Bool {
  require(boolean_expression == true, "Error message");
}
```

Use instead:

```tact
fun foo(x : Int): Bool {
  require(boolean_expression, "Error message");
}
```

### PR DESCRIPTION
**Pull request by Esorat**
Closes #92 

- [x] I have updated `CHANGELOG.md`
- [x] I have added tests to demonstrate the contribution is correctly implemented
- [x] No test failures were reported when running `yarn test-all`
- [x] I did not do unrelated and/or undiscussed refactorings

### CODE REVIEW

File: src/detectors/builtin/etaLikeSimplifications.ts [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+      }
+    }
+  }
+
+  private isBooleanLiteral(
+    expr: AstExpression | null | undefined,
+    value?: boolean,
+  ): boolean {
+    if (!expr) return false;
+    if (expr.kind === "boolean") {
```
Comment (jubnzv) [inline src/detectors/builtin/etaLikeSimplifications.ts:153]: Let's use the constant evaluator

File: src/detectors/builtin/etaLikeSimplifications.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  private getSimplifiedBooleanExpression(binaryExpr: AstOpBinary): string {
+    const exprText = (expr: AstExpression): string => {
+      if (expr.kind === "id") {
+        return idText(expr);
+      }
+      return "expression";
+    };
+
+    if (this.isBooleanLiteral(binaryExpr.right, true)) {
+      if (binaryExpr.op === "==") {
```
Comment (jubnzv) [inline src/detectors/builtin/etaLikeSimplifications.ts:169]: I would simplify creating a constant here: `const negated = binaryExpr.op === "=="` and using it in this function.

File: src/detectors/builtin/etaLikeSimplifications.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+      }
+    }
+    return "expression";
+  }
+
+  private getConditionText(expr: AstExpression): string {
+    if (expr.kind === "id") {
+      return idText(expr);
+    }
+    return "condition";
```
Comment (jubnzv) [inline src/detectors/builtin/etaLikeSimplifications.ts:188]: What does it mean?

File: src/detectors/builtin/etaLikeSimplifications.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        if (
+          this.isBooleanLiteral(trueReturn.expression, true) &&
+          this.isBooleanLiteral(falseReturn.expression, false)
+        ) {
+          warnings.push(
+            this.makeWarning(
+              "Simplify 'if' statement by returning the condition directly",
+              stmt.loc,
+              {
+                suggestion: "Replace with 'return condition;'",
```
Comment (jubnzv) [inline src/detectors/builtin/etaLikeSimplifications.ts:96]: Pretty print it?

**Review by jubnzv — CHANGES_REQUESTED** [review]
The core logic seems correct. Let's adjust the implementation.

**Comment by jubnzv** [comment]
We should also merge the master in this branch and resolve conflicts, including those in the changelog.

File: src/detectors/builtin/etaLikeSimplifications.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+              )}' instead`,
+            },
+          ),
+        );
+      }
+    }
+  }
+
+  private isBooleanLiteral(
+    expr: AstExpression | null | undefined,
```
[the remark this answers is not in the document]
Comment (jubnzv) [reply src/detectors/builtin/etaLikeSimplifications.ts:148]: I'm not sure why it is resolved without any changes.

[the remark this answers is not in the document]
Comment (Esorat) [reply src/detectors/builtin/etaLikeSimplifications.ts:148]: I included both null and undefined in parameter type. This allows function to handle both cases without requiring type assertions or conversions at the call site.

File: CHANGELOG.md [hunk 6]
```diff
@@ -8,13 +8,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 ### Added
+- `EtaLikeSimplifications` detector: PR [#198](https://github.com/nowarp/misti/pull/198)
 - `ShortCircuitCondition` detector: PR [#202](https://github.com/nowarp/misti/pull/202)
 ### Changed 
 - `SuspiciousMessageMode` detector now suggests using SendDefaultMode instead of 0 for mode: PR [#199](https://github.com/nowarp/misti/pull/199/)
 
 ## [0.5.0] - 2024-10-31
 
 ### Added
+- `EtaLikeSimplifications` detector: PR [#198](https://github.com/nowarp/misti/pull/198)
```
Comment (jubnzv) [inline CHANGELOG.md:19]:
```suggestion
```

**Review by jubnzv — APPROVED** [review]
LGTM

**Pull request merged**